088-rebalance-progress-status

```diff
@@ -0,0 +1,328 @@
+  "numPendingPartitionMovements": 20,
+  "numTotalPartitionMovements": 20,
+  "state": "INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
+  "totalDataToMove": 0,
+  "triggeredSelfHealingTaskId": "",
+  "triggeredTaskReason": "No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
+  "triggeredUserTaskId": "0230d401-6a36-430e-9858-fac8f2edde93"
+}
+```
+
+For determining which fields are included for different executor states e.g. NO_TASK_IN_PROGRESS, STARTING_EXECUTION, INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS whether the verbose parameter is provided or not, refer to the code [here](https://github.com/linkedin/cruise-control/blob/2.5.141/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorState.java#L427)
+
+For an exhaustive list of all the fields that could be included, refer to Cruise Control’s OpenAPI spec, refer to the file [here](https://github.com/linkedin/cruise-control/blob/2.5.141/cruise-control/src/main/resources/yaml/responses/executorState.yaml)
+
+The “non-verbose” JSON payload from the ExecutorState is already too verbose to include in the `KafkaRebalance` status in its entirety. 
+However, having the information available to users is still useful especially when debugging the state of a partition rebalance. 
+Therefore, we will store the JSON payload in its own ConfigMap, “rebalanceProgressConfigMap”. 
+For this initial feature enhancement we will only store the “non-verbose” JSON output but we will still have a good amount of space remaining in the ConfigMap should we make the verbosity configurable in the future. 
+
+The ConfigMap with ExecutorStatus included would section:
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-rebalance-progress
+  …
+data:
+  executorState: [1]
+`{"abortingPartitions":0,"averageConcurrentPartitionMovementsPerBroker":5,"finishedDataMovement":0,"maximumConcurrentPartitionMovementsPerBroker":5,"minimumConcurrentPartitionMovementsPerBroker":5,"numFinishedPartitionMovements":0,"numInProgressPartitionMovements":0,"numPendingPartitionMovements":20,"numTotalPartitionMovements":20,"state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS","totalDataToMove":0,"triggeredSelfHealingTaskId":"","triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)","triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"}
+`
+```
+[1] An example of the ExecutorState JSON of an inter-broker partition rebalance. 
+
+### Progress Update Cadence
+
+For ease of implementation and minimizing the load on the CruiseControl REST API server, we would only query the CruiseControlState endpoint and update the “progress” section upon `KafkaRebalance` resource reconciliation. 
+The progress section will never be more out of date longer than the reconciliation period and even if the rebalance runs into an error or “NotReady” state, the “progress” section would still be updated on that KafkaRebalance resource reconciliation along with any error.
```

Comment:
How do you avoid tight reconciliation loop as update to the status will trigger new reconciliation that will update the status, trigger new reconciliation etc.?

Reply:
That is a very good point. Maybe we need to post a timestamp of last progress check and if it is less than the reconciliation period then skip?

Reply:
The general rule is to not include things like that in the status. Using some timestamp for that would probably need to be handled when getting the progress data and not when updating the status, as that is a shared code and it might be complicated to put it there.

Reply:
> That is a very good point. Maybe we need to post a timestamp of last progress check and if it is less than the reconciliation period then skip?

>The general rule is to not include things like that in the status. Using some timestamp for that would probably need to be handled when getting the progress data and not when updating the status, as that is a shared code and it might be complicated to put it there.

I should be able to use the existing timestamp in `metadata.managedFields` field of `KafkaRebalance`  resource to know when the resource was last updated, then only update the progress section if that timestamp is older than the reconciliation period.

Reply:
Do you really know what changes are tracked there and whose timestamp it migth have?

Reply:
> The progress section will never be more out of date longer than the reconciliation period this part might not be true always. For example, if CC REST API returned an error for some reason and the executor state could not retrieved, would we wait for the next reconciliation to retry? In which case, the progress section would be out of date.

That is a good point. Although the KafkaRebalanceOperator would report the CC REST API error in the `KafkaRebalance` resource status, it would not have the information it needed to update the progress section. In the event that the executor state could not retrieved, the KafkaRebalanceOperator could update the values of the progress section with some placeholder value to indicate retrieval failure (e.g. null, N/A, "Information could not be retrieved") and update the progress section next reconciliation. What are your thoughts on that @tinaselenge?

Reply:
Yes, I think having some indication/placeholder as you suggested when executor state could not retrieved would be helpful to the user because KafkaRebalance resource status might not be always clear if it had any impact on the progress section or not.

Reply:
I don't think you should rely on `metadata.managedFields[].time` as those are internal Kafka fields with completely different purposes.

Reply:
> I don't think you should rely on metadata.managedFields[].time as those are internal Kafka fields with completely different purposes.

If we are not allowed to maintain a timestamp in the progress section specifying when it was last changed or rely on the `metadata.managedFields[].time` field of the custom resource then we will either have to find another way of tracking when the resource was last updated or try another approach for preventing tight reconciliation loops.

I'll see what I come up with and get back to you.

Reply:
I'm not sure you cannot have a timestamp in the status. The question is how you work with the timestamp, how you use it, and when/how you update it. But in general, the easiest solution is to store the progress in a config map which you can simply update in very reconciliation and as you don't watch you do not need to b worried about what it triggers. Event might be other option for the progress tracking maybe? I do not like them very much and I think they are pretty useless for tracking the restart events. But if you publish the events to the KafkaRebaance resource, it might be more useful than for Pods.

Reply:
> I'm not sure you cannot have a timestamp in the status. The question is how you work with the timestamp, how you use it, and when/how you update it. 

If we were to maintain a timestamp in the status we would add it with the `progress` section upon initial creation, then update it on the next reconciliation when its value was older than reconciliation period. With a timestamp of when the resource was last changed, we could easily avoid triggering unwanted reconciliations. 

> But in general, the easiest solution is to store the progress in a config map which you can simply update in very reconciliation and as you don't watch you do not need to b worried about what it triggers.

This is a really interesting idea. TBH I hadn't thought of storing the progress information in a `ConfigMap` instead of the `KafkaRebalance` status. We were planning on maintaining a `ConfigMap` for executor state information anyway and yes, in this way we could avoid triggering reconciliations upon `progress` updates. We would still need to add a `progress` section with a reference to the `ConfigMap` but this would only need to be added/removed once per state change. 

Although maintaining the `progress` information in the `ConfigMap` would be the simplest solution, I still feel that the UX of maintaining the `progress` information in the `KafkaRebalance` status would still be worth the added implementation complexity. Any thoughts on this @ppatierno @tomncooper ?

> Event might be other option for the progress tracking maybe? I do not like them very much and I think they are pretty useless for tracking the restart events. But if you publish the events to the KafkaRebaance resource, it might be more useful than for Pods.

I hadn't thought of using events either but I'll think more on this. My only concern for storing the  `progress` in the events would be the UX of getting the progress information, the initial idea was that the `progress` information would be easily found and read by users in the `KafkaRebalance` resource.

Reply:
>If we were to maintain a timestamp in the status we would add it with the progress section upon initial creation, then update it on the next reconciliation when its value was older than reconciliation period. With a timestamp of when the progress was last changed, we could easily avoid triggering unwanted reconciliations.

Yes, you would need some custom logic such as `if the timestamp is older than X minutes, update the progress. If not, just reuse the old progress`.

> I hadn't thought of using events either but I'll think more on this. My only concern for storing the progress in the events would be the UX of getting the progress information, the initial idea was that the progress information would be easily found and read by users in the KafkaRebalance resource.

`kubectl describe kr` should show you the events I think. Also most UIs would normally show the events when you list the custom resource.

Reply:
Discussed w/ Paolo and Tom, they agreed storing the `progress` information in the `ConfigMap` would simplify the implementation and that doing so wouldn't significantly change the UX. Given the executor state information is already going to be stored in the `ConfigMap` it probably makes the most sense to maintain our progress information there as well. Let me update the proposal to show what it would look like